fineract-provider/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java

```diff
@@ -69,45 +76,43 @@ public class SynchronousCommandProcessingService implements CommandProcessingSer
     private final IdempotencyKeyGenerator idempotencyKeyGenerator;
     private final FineractProperties fineractProperties;
 
+    private final InitialCommandSaver initialCommandSaver;
+
     @Transactional
     @Override
     public CommandProcessingResult executeCommand(final CommandWrapper wrapper, final JsonCommand command,
             final boolean isApprovedByChecker) {
 
         final boolean rollbackTransaction = configurationDomainService.isMakerCheckerEnabledForTask(wrapper.taskPermissionName());
+        String idempotencyKey = getIdempotencyKey(wrapper);
+        CommandSource existingCommand = commandSourceRepository.findByActionNameAndEntityNameAndIdempotencyKey(wrapper.actionName(),
+                wrapper.entityName(), idempotencyKey);
+        final CommandProcessingResult result;
+        if (existingCommand != null) {
+            if (UNDER_PROCESSING.getValue().equals(existingCommand.getStatus())) {
+                throw new DuplicateCommandException(wrapper.actionName(), wrapper.entityName(), wrapper.getIdempotencyKey(),
+                        wrapper.getJson(), UNDER_PROCESSING.getValue());
+            } else if (ERROR.getValue().equals(existingCommand.getStatus())) {
+                throw new DuplicateCommandException(wrapper.actionName(), wrapper.entityName(), wrapper.getIdempotencyKey(),
+                        existingCommand.getResult(), ERROR.getValue());
+            } else if (PROCESSED.getValue().equals(existingCommand.getStatus())) {
+                throw new DuplicateCommandException(wrapper.actionName(), wrapper.entityName(), wrapper.getIdempotencyKey(),
+                        existingCommand.getResult(), PROCESSED.getValue());
+            }
+        }
 
-        final NewCommandSourceHandler handler = findCommandHandler(wrapper);
+        final AppUser maker = context.authenticatedUser(wrapper);
+        CommandSource commandSourceResult = getInitialCommandSource(wrapper, command, maker, idempotencyKey);
+        initialCommandSaver.save(commandSourceResult);
```

Comment:
Couldn't we make this separation better? Just for creating the InitialCommandSaver, to work with the declarative transaction, it's fishy.

Instead why don't we move the whole initial command creation logic into there (including the getInitialCommandSource method).

That way the implementation would be more robust by just calling a "initialCommandCreator.create" method which would create and save the new command. 

Thoughts?

Or... you could create a separate repository method which always saves the entity in a new transaction (using the Custom Spring repository way for example) and then you don't need to have this InitialCommandSaver.

Reply:
will do

Reply:
> because transaction forcing inside transactions only works with public method calls
That's only true for declarative tx management. If you use a TransactionTemplate, you're pretty much good to go.

Reply:
The codebase is already using declarative tx mangement at numerous places. (very appropriately instead of mixing techniques)